[TorchTidy] Add Pattern to detect Synchronous Data Loader

davidchencsl opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82261
* #82248
* #81941
* #81921
* __->__ #81740

Summary: By setting num_workers > 0 in DataLoader, we can achieve async data loading, which is non blocking to the computation. This helps speed up the training process. By matching the call structure, we can detect if we are using Synchronous Data Loader.

Test Plan:
Added test in test.profiler.py

Differential Revision: [D38082644](https://our.internmc.facebook.com/intern/diff/D38082644)

davidchencsl committed:
[TorchTidy] Add Pattern to detect Synchronous Data Loader

[ghstack-poisoned]

davidchencsl committed:
Update on "[TorchTidy] Add Pattern to detect Synchronous Data Loader"




[ghstack-poisoned]

davidchencsl committed:
Update on "[TorchTidy] Add Pattern to detect Synchronous Data Loader"




[ghstack-poisoned]

robieta commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -263,7 +264,7 @@ def skip(self):
         has_tf32 = all(
             int(arch[3:]) >= 80 for arch in torch.cuda.get_arch_list())
-        return has_tf32 is False
+        return has_tf32 is False or super().skip or not self.prof.record_shapes
```
Looks like there are several updates to prior PRs that you discovered here. You should move them to their respective PRs.

robieta commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -345,6 +345,58 @@ def match(self, event: _ProfilerEvent):
+        )
+        # We precompile the regex to avoid recompiling it every time we match
+        # The four back slashes are to escape windows path separator
+        seperator = "\\\\" if os.sep == "\\" else os.sep
```
If you steal the binding from `https://github.com/pytorch/pytorch/pull/81895` then you won't have to mess with regex; you'll be able to directly access the underlying fields. (And you can just compare the filename with `os.path.join(["torch", "utils", "data", "dataloader.py])` to get the right path format.

davidchencsl replied:
One advantage for me to use regex here is that I can precompile the pattern as regex object. I don't need to process all the event name. Versus applying special treatment for all the event name

robieta commented on torch/profiler/_pattern_matcher.py:
```diff
@@ -345,6 +345,58 @@ def match(self, event: _ProfilerEvent):
+            r"torch/utils/data/dataloader.py\([0-9]+\): check_worker_number_rationality$"
+            .replace("/", seperator))
+
+    def match(self, event: _ProfilerEvent):
```
This could use some comments to explain your reasoning. I think it also needs some sort of check for "waiting". If my dataloader is able to keep up with training (maybe I already have the data in memory, maybe I'm handling parallelism elsewhere, maybe it's just really cheap) then `num_workers=0` is appropriate.

robieta replied:
(You'll have to override this for test stability, of course.)

davidchencsl replied:
I think I should add a TODO here. I was planning to add a check to the idle time during data loading, but I want to do it after kineto type is in.

davidchencsl committed:
Update on "[TorchTidy] Add Pattern to detect Synchronous Data Loader"


Summary: By setting num_workers > 0 in DataLoader, we can achieve async data loading, which is non blocking to the computation. This helps speed up the training process. By matching the call structure, we can detect if we are using Synchronous Data Loader.

Test Plan:
Added test in test.profiler.py

[ghstack-poisoned]

davidchencsl committed:
Update on "[TorchTidy] Add Pattern to detect Synchronous Data Loader"


Summary: By setting num_workers > 0 in DataLoader, we can achieve async data loading, which is non blocking to the computation. This helps speed up the training process. By matching the call structure, we can detect if we are using Synchronous Data Loader.

Test Plan:
Added test in test.profiler.py

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

davidchencsl committed:
Update on "[TorchTidy] Add Pattern to detect Synchronous Data Loader"


Summary: By setting num_workers > 0 in DataLoader, we can achieve async data loading, which is non blocking to the computation. This helps speed up the training process. By matching the call structure, we can detect if we are using Synchronous Data Loader.

Test Plan:
Added test in test.profiler.py

Differential Revision: [D38082644](https://our.internmc.facebook.com/intern/diff/D38082644)

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

davidchencsl committed:
Update on "[TorchTidy] Add Pattern to detect Synchronous Data Loader"


Summary: By setting num_workers > 0 in DataLoader, we can achieve async data loading, which is non blocking to the computation. This helps speed up the training process. By matching the call structure, we can detect if we are using Synchronous Data Loader.

Test Plan:
Added test in test.profiler.py

Differential Revision: [D38082644](https://our.internmc.facebook.com/intern/diff/D38082644)

[ghstack-poisoned]

davidchencsl commented:
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

robieta reviewed (approved):
LGTM

davidchencsl commented:
@pytorchbot merge
